apiserver/common/modeluser.go

```diff
@@ -13,23 +13,22 @@ import (
 	coremodel "github.com/juju/juju/core/model"
 	"github.com/juju/juju/core/permission"
 	"github.com/juju/juju/core/user"
-	"github.com/juju/juju/domain/access"
 	"github.com/juju/juju/rpc/params"
 )
 
-type accessService interface {
-	// GetModelUsers will retrieve basic information about all users with
-	// permissions on the given model UUID.
+type modelService interface {
+	// GetModelUsers will retrieve basic information about users with permissions on
+	// the given model UUID.
+	// If filterUsers is empty, all model users will be returned. Otherwise, only
+	// the users in filterUsers will be returned.
 	// If the model cannot be found it will return modelerrors.NotFound.
-	// If no permissions can be found on the model it will return
-	// accesserrors.PermissionNotValid.
-	GetModelUsers(ctx context.Context, apiUser user.Name, modelUUID coremodel.UUID) ([]access.ModelUserInfo, error)
+	GetModelUsers(ctx context.Context, modelUUID coremodel.UUID, filterUsers []user.Name) ([]coremodel.ModelUserInfo, error)
 }
 
-// ModelUserInfo gets model user info from the accessService and converts it
+// ModelUserInfo gets model user info from the modelService and converts it
 // into params.ModelUserInfo.
-func ModelUserInfo(ctx context.Context, service accessService, apiUser names.UserTag, modelTag names.ModelTag) ([]params.ModelUserInfo, error) {
-	userInfo, err := service.GetModelUsers(ctx, user.NameFromTag(apiUser), coremodel.UUID(modelTag.Id()))
+func ModelUserInfo(ctx context.Context, service modelService, modelTag names.ModelTag, filter []user.Name) ([]params.ModelUserInfo, error) {
```

Comment:
To me, this `filter` feels a bit overkill if the only ever options are going to be all users/no filter or a single user. In the second case it's not really a "list all users" call anymore, it's an existence check for a single user. In which case we should probably just have a `GetModelUser` function or similar which retrieves info for a single user, and handle the notfound error.

Reply:
If you're gonna stick with the filter, maybe push it down into the state layer to get the efficiency benefits

Reply:
**thought:** Which bother me more is that filter is not an actual filter, but a list of name, and I have to infer from my own knowledge of "how think programmer in the 21th century" to get to an expected behavior:

* If empty return all
* if not empty, return  only modelUserInfo whose Name field is somewhat contained in the filter one.

I think we are leaking implementation details (I assumes that deep inside, we pass the list to the DB to a `WHERE` close. It is why, in this interface, we pass a list. From my point of view, It means that we are leaking implementation details from the DB to the API.

My two cents ;)

Reply:
Both these points are fair, I think I'll change it to do as @barrettj12 says, have a separate `GetModelUser` function. I had thought it might be good to allow more flexibility but in retrospect the filter is just adding additional complexity for no reason.

I'm very tempted to just have the `GetModelUser` method on the service, and not on the state, i.e. just find the single user requested in the list of the results. This would reduce code complexity and the method is not on a heavy call path and adding a second function or conditional logic in the SQL to just select a single user seems like it wouldn't give much benefit, if anyone is in strong opposition to this though I'm happy to implement it.

Reply:
> I'm very tempted to just have the `GetModelUser` method on the service, and not on the state, i.e. just find the single user requested in the list of the results.

That seems fine to me as an internal implementation detail, you'll just have to ensure NotFound errors are coming through.